### ISSUE TITLE
Problem with Select Projection

### ISSUE THREAD

**Issue by VincentDoreau13** [rgvlee/EntityFrameworkCore.Testing#91]
Hi,

Thanks for the tool on github but I have a problem with using the tool.

I am using a Select in my entity framework query with a projection to a file.

It works fine except the part where there are properties going around in my entities. However, this does not pose a problem in the production of my application.

I tried to reproduce the problem on your test part but I did not succeed. If I have to I will keep trying.

Here is the error I get during my test: 

```
System.InvalidCastException : Unable to cast object of type 'System.Linq.Expressions.NewExpression' to type 'System.Linq.Expressions.MethodCallExpression'.
[xUnit.net 00:00:02.14]       Stack Trace:
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryQueryExpression.AddSubqueryProjection(ShapedQueryExpression shapedQueryExpression, Expression& innerShaper)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.Visit(Expression expression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.VisitMemberAssignment(MemberAssignment memberAssignment)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.VisitMemberInit(MemberInitExpression memberInitExpression)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.Visit(Expression expression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.VisitMemberAssignment(MemberAssignment memberAssignment)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.VisitMemberInit(MemberInitExpression memberInitExpression)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.Visit(Expression expression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryProjectionBindingExpressionVisitor.Translate(InMemoryQueryExpression queryExpression, Expression expression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryQueryableMethodTranslatingExpressionVisitor.TranslateSelect(ShapedQueryExpression source, LambdaExpression selector)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.InMemory.Query.Internal.InMemoryQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
[xUnit.net 00:00:02.14]            at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
[xUnit.net 00:00:02.14]            at EntityFrameworkCore.Testing.Common.AsyncEnumerator`1..ctor(IEnumerable`1 enumerable)
[xUnit.net 00:00:02.14]            at EntityFrameworkCore.Testing.Common.AsyncEnumerable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
[xUnit.net 00:00:02.14]            at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
[xUnit.net 00:00:02.14]            at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)

**Comment by rgvlee** [comment]
Thanks for reporting the issue.
Moq or NSubstitute? Version number?
A minimum reproducible example will help a lot if you can provide one as I don't know if I have enough information to work with at the moment.

**Comment by VincentDoreau13** [comment]
I am using EntityFrameworkCore.Testing.Moq version 3.0.0.
Sorry I had failed to properly identify the problem and reproduce it correctly in your tests.
I finally succeeded and I will be able to detail this for you.

Here is the test that I added in your tests (BaseForQueryableTests) :

```CSharp
public class TestEntityProjection
{
    public int Test { get; set; }

    public IEnumerable<int> Test1 { get; set; }
}

public static Expression<Func<T, TestEntityProjection>> Projection => entity => new TestEntityProjection()
{
    Test = entity.SecondBaseTests.Sum(x => x.Id),
    Test1 = entity.ThirdBaseTests.Select(x => x.Id),
};

[Test]
public virtual void SelectWithMethodCall_ReturnsSequence()
{
    SeedQueryableSource();

    var actualResult1 = Queryable.Select(Projection).ToList();
}
```

And here's BaseTestEntity for the test:

```CSharp
public abstract class BaseTestEntity
{
    public Guid Id { get; set; }

    public string FullName { get; set; }

    public decimal Weight { get; set; }

    public decimal Height { get; set; }

    public DateTime DateOfBirth { get; set; }

    public DateTime CreatedAt { get; set; }

    public DateTime LastModifiedAt { get; set; }

    public IList<SecondBaseTest> SecondBaseTests { get; set; }

    public IList<ThirdBaseTest> ThirdBaseTests { get; set; }
}

public class SecondBaseTest
{
    public int Id { get; set; }
}

public class ThirdBaseTest
{
    public int Id { get; set; }
}
```

This fails for ByPropertyDbTests and ByTypeDbSetTests.

If I have a Sum() and a Select() it doesn't work. But if I comment out one of the two lines then it works, the same if I put everything in Sum() or everything in Select().

Thank you in advance for your help.

**Comment by rgvlee** [comment]
I've done some preliminary testing and it looks like the same test fails for the in-memory database provider. That'd make sense as I believe that it ends up invoking `CreateQuery` on the queryable provider, and in this instance that'll be falling back onto the one provided by the in-memory database provider. This also lines up with the stack trace.

The same operation would probably work fine for a read-only db set.

It does work fine on an ordinary `List`. To take this further I'll need to confirm that this behaviour is permitted by the SQL Server database provider as that's the primary scope of support for the library.

**Comment by rgvlee** [comment]
This [branch](https://github.com/rgvlee/EntityFrameworkCore.Testing/compare/bugfix/3.x/Issue-91_ResolveSourceBeforeCreateQuery) includes a fix for this issue, along with showing that the in-memory database provider suffers the same issue.

I did notice that the issue goes away if the projection is a `Func<TDataEntity, TProjectionEntity>` rather than a `Expression<Func<TDataEntity, TProjectionEntity>>`.

I have released 3.0.4-beta01 which includes this fix. Once I have either a test or have proven the same issue exists with the SQL Server database provider it will be included in a release.
